src/Orleans/IDs/Interner.cs

```diff
@@ -96,31 +96,34 @@ public Interner(int initialSize, TimeSpan cleanupFreq)
         /// <param name="key">key to find</param>
         /// <param name="creatorFunc">function to create new object and store for this key if no cached copy exists</param>
         /// <returns>Object with specified key - either previous cached copy or newly created</returns>
-        public T FindOrCreate(K key, Func<T> creatorFunc)
+        public T FindOrCreate(K key, Func<K, T> creatorFunc)
         {
-            T obj = null;
-            WeakReference cacheEntry = internCache.GetOrAdd(key,
-                (k) =>
-                {
-                    obj = creatorFunc();
-                    return new WeakReference(obj);
-                });
-            if (cacheEntry != null)
+            T result;
+            WeakReference<T> cacheEntry;
+
+            // Attempt to get the existing value from cache.
+            internCache.TryGetValue(key, out cacheEntry);
+
+            // If no cache entry exists, create and insert a new one using the creator function.
+            if (cacheEntry == null)
```

Comment:
@sergeybykov the copy is made before the call. Note that the signature of the function is 
```C#
private bool TryAddInternal(TKey key, TValue value, bool updateIfExists, bool acquireLock, out TValue resultingValue)
```

The call is made to obtain `value` prior to invoking `TryAddInternal`